flash/partition_info/partition_info.c

```diff
@@ -0,0 +1,164 @@
+#define WORD_PADDED_BYTES(x)  ((x + 3) & ~3)
+
+#define WORD_SIZE  4
+#define PARTITION_LOCATION_AND_FLAGS_SIZE  2
+#define PARTITION_ID_SIZE                  2
+#define PARTITION_NAME_MAX                 127
+#define PARTITION_TABLE_FIXED_INFO_SIZE    (4 + PARTITION_TABLE_MAX_PARTITIONS * (PARTITION_LOCATION_AND_FLAGS_SIZE + PARTITION_ID_SIZE))
+
+/*
+ * Stores partition table information and data read status
+ */
+typedef struct {
+    uint32_t table[PARTITION_TABLE_FIXED_INFO_SIZE];
+    uint32_t fields;
+    int partition_count;
+    uint32_t unpartitioned_space_first_sector;
+    uint32_t unpartitioned_space_last_sector;
+    uint32_t permission;
+    int current_partition;
+    size_t idx;
+    int status;
+} pico_partition_table_t;
+
+/*
+ * Stores information on each partition
+ */
+typedef struct {
+    uint32_t first_sector;
+    uint32_t last_sector;
+    uint32_t flags_and_permissions;
+    uint64_t partition_id;
+    char name[PARTITION_NAME_MAX + 1];  // name length is indicated by 7 bits
+} pico_partition_t;
+
+
+/*
+ * Read the partition table information.
+ *
+ * See the RP2350 datasheet 5.1.2, 5.4.8.16 for flags and structures that can be specified.
+ */
+int pico_partitions_open(pico_partition_table_t *pt) {
+    // Reads a fixed size fields
+    int rc = rom_get_partition_table_info(pt->table, sizeof(pt->table),
+                                          (PT_INFO_PT_INFO |
+                                           PT_INFO_PARTITION_LOCATION_AND_FLAGS |
+                                           PT_INFO_PARTITION_ID));
+    if (rc < 0) {
+        pt->status = rc;
+        return rc;
+    }
+
+    size_t pos = 0;
+    pt->fields = pt->table[pos++];
+    pt->partition_count = pt->table[pos++] & 0x000000FF;
+    uint32_t location = pt->table[pos++];
+    pt->unpartitioned_space_first_sector = PART_LOC_FIRST(location);
+    pt->unpartitioned_space_last_sector = PART_LOC_LAST(location);
+    pt->permission = pt->table[pos++];
+    pt->current_partition = 0;
+    pt->idx = pos;
+    pt->status = 0;
+
+    return 0;
+}
+
+/*
+ * Extract each partition information
+ */
+bool pico_partitions_next(pico_partition_table_t *pt, pico_partition_t *p) {
+    if (pt->current_partition >= pt->partition_count) {
+        return false;
+    }
+
+    size_t idx = pt->idx;
+    uint32_t location = pt->table[idx++];
+    p->first_sector = PART_LOC_FIRST(location);
+    p->last_sector = PART_LOC_LAST(location);
+    p->flags_and_permissions = pt->table[idx++];
+
+    if (p->flags_and_permissions & PICOBIN_PARTITION_FLAGS_HAS_ID_BITS) {
+        uint32_t id_low  = pt->table[idx++];
+        uint32_t id_high = pt->table[idx++];
+        p->partition_id = ((uint64_t)id_high << 32) | id_low;
+    } else {
+        p->partition_id = 0;
+    }
+    pt->idx = idx;
+
+    if (p->flags_and_permissions & PICOBIN_PARTITION_FLAGS_HAS_NAME_BITS) {
+        // Read variable length fields
+        uint32_t name_buf[(PARTITION_NAME_MAX + 1) / WORD_SIZE] = {0};
```

Comment:
I don't think `WORD_SIZE` needs to be defined, you could use `sizeof(uint32_t)` instead and remove the `WORD_SIZE` define at the top

```suggestion
        uint32_t name_buf[(PARTITION_NAME_MAX + 1) / sizeof(uint32_t)] = {0};
```